vertical-pod-autoscaler/pkg/admission-controller/logic/server_test.go

```diff
@@ -0,0 +1,284 @@
+package logic
+
+import (
+	"encoding/json"
+	"fmt"
+	apiv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	vpa_api_util "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/utils/vpa"
+	"strings"
+	"testing"
+)
+
+const (
+	cpu        = "cpu"
+	unobtanium = "unobtanium"
+)
+
+type fakePodPreProcessor struct {
+	e error
+}
+
+func (fpp *fakePodPreProcessor) Process(pod apiv1.Pod) (apiv1.Pod, error) {
+	return pod, fpp.e
+}
+
+type fakeRecommendationProvider struct {
+	resources              []ContainerResources
+	containerToAnnotations vpa_api_util.ContainerToAnnotationsMap
+	name                   string
+	e                      error
+}
+
+func (frp *fakeRecommendationProvider) GetContainersResourcesForPod(pod *apiv1.Pod) ([]ContainerResources, vpa_api_util.ContainerToAnnotationsMap, string, error) {
+	return frp.resources, frp.containerToAnnotations, frp.name, frp.e
+}
+
+func addResourcesPatch(idx int) patchRecord {
+	return patchRecord{
+		"add",
+		fmt.Sprintf("/spec/containers/%d/resources", idx),
+		apiv1.ResourceRequirements{},
+	}
+}
+
+func addRequestsPatch(idx int) patchRecord {
+	return patchRecord{
+		"add",
+		fmt.Sprintf("/spec/containers/%d/resources/requests", idx),
+		apiv1.ResourceList{},
+	}
+}
+
+func addResourceRequestPatch(index int, res, amount string) patchRecord {
+	return patchRecord{
+		"add",
+		fmt.Sprintf("/spec/containers/%d/resources/requests/%s", index, res),
+		resource.MustParse(amount),
+	}
+}
+
+func addAnnotationRequest(updateResources [][]string) patchRecord {
+	requests := make([]string, 0)
+	for idx, podResources := range updateResources {
+		podRequests := make([]string, 0)
+		for _, resource := range podResources {
+			podRequests = append(podRequests, resource+" request")
+		}
+		requests = append(requests, fmt.Sprintf("container %d: %s", idx, strings.Join(podRequests, ", ")))
+	}
+
+	vpaUpdates := fmt.Sprintf("Pod resources updated by name: %s", strings.Join(requests, "; "))
+	return patchRecord{
+		"add",
+		"/metadata/annotations",
+		map[string]string{
+			"vpaUpdates": vpaUpdates,
+		},
+	}
+}
+
+func eqPatch(a, b patchRecord) bool {
+	aJson, aErr := json.Marshal(a)
+	bJson, bErr := json.Marshal(b)
+	return string(aJson) == string(bJson) && aErr == bErr
+}
+
+func TestGetPatchesForResourceRequest(t *testing.T) {
+	tests := []struct {
+		name                 string
+		podJson              []byte
+		namespace            string
+		preProcessorError    error
+		recommendResources   []ContainerResources
+		recommendAnnotations vpa_api_util.ContainerToAnnotationsMap
+		recommendName        string
+		recommendError       error
+		expectPatches        []patchRecord
+		expectError          error
+	}{
+		{
+			"invalid JSON",
```

Comment:
This might be nitpicking too much, but I think the cases would be more readable with the field names explicitly given.

Reply:
Done. IDE shows field names but spelling them out makes them visible in code reviews. Also I get to skip some nils.